arch/sim/include/types.h

```diff
@@ -69,13 +69,41 @@ typedef unsigned char      _uint8_t;
 typedef signed short       _int16_t;
 typedef unsigned short     _uint16_t;
 
+#if defined(__INT32_TYPE__)
+typedef __INT32_TYPE__     _int32_t;
+#else
 typedef signed int         _int32_t;
+#endif
+
+#if defined(__UINT32_TYPE__)
+typedef __UINT32_TYPE__    _uint32_t;
+#else
 typedef unsigned int       _uint32_t;
+#endif
 
+#if defined(__INT64_TYPE__)
+typedef __INT64_TYPE__     _int64_t;
+#else
 typedef signed long long   _int64_t;
+#endif
+
+#if defined(__UINT64_TYPE__)
+typedef __UINT64_TYPE__    _uint64_t;
+#else
 typedef unsigned long long _uint64_t;
+#endif
 #define __INT64_DEFINED
 
+#if defined(__INTMAX_TYPE__)
+typedef __INTMAX_TYPE__   _intmax_t;
+#define __INTMAX_DEFINED
```

Comment:
how about we also typedef _intmax_t in __INTMAX_TYPE__ undefined case?

Reply:
i guess the int64 fallback in stdint.h is enough.

Reply:
but why not to make _intmax_t similar as other type? anyway, you need define _intmax_t in arch.h now.

Reply:
> but why not to make _intmax_t similar as other type?

just because i didn't need to change it.

> anyway, you need define _intmax_t in arch.h now.

can you explain?
in my patch, _intmax_t is used only when __INTMAX_DEFINED is defined.

Reply:
I mean that intmax_t just need appear in stdint.h before your patch, it's reasonable to different from other basic type. But, now we have to typedef _intmax_t in arch.h just like other basic type, why not use the same approach to always typedef _intmax_t?

Reply:
we can. i just wanted to make the change small. do you want me to go through all archs in this PR?

Reply:
Yes, I think the consistent style and equally support all arch is NuttX's unique characteristics.

Reply:
ok. i will take a look later.

Reply:
i updated all types.h except arch/z80/include/chip/types.h.
the arch/z80/include/chip directory seems like a stale copy of arch/z80/include/z8.
@patacongo should arch/z80/include/chip/types.h be identical to arch/z80/include/z8/types.h?

Reply:
Thanks @yamt, this patch will make we found the wrong usage of printf/scanf more easiler.
And syslog and debug.h could improve too.

Reply:
after thinking a bit more, this approach needs to use __UINT32_FMTu__ for PRIu32 etc to be consistent.

Reply:
Yes, you are right.

Reply:
unfortunately it seems that `__UINT32_FMTu__` style macros are clang-only.

i guess we just need to change the "default" types in arch types.h and inttypes.h to match what gcc expects.
for some archs, they are different right now.
eg. xtensa-esp32-elf-gcc has `#define __INT32_TYPE__ int` while we have signed long https://github.com/apache/incubator-nuttx/blob/713a21e57c27db7bddbb1e841714829d62d5ef7d/arch/xtensa/include/types.h#L72

Reply:
i'm thinking to make arch types.h like the following
so that it produces an explicit error if the compiler expectation (`__UINTxx_TYPE__`) doesn't match with ours.

```
typedef unsigned int _uint32_t;
#if defined(__UINT32_TYPE__)
typedef __UINT32_TYPE__ _uint32_t;
#endif
```

the C standard allows typedef redefinitions as far as the type is same.
(n1570.pdf 6.7/3)

Reply:
why not directly typedef __UINT32_TYPE__ _uint32_t?

Reply:
because some compilers might not define `__UINT32_TYPE__`.

Reply:
In that case, I suppose print_format also doesn't support too.

Reply:
a quick googling suggested: https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2010/ms235402(v=vs.100)?redirectedfrom=MSDN#advanced-annotations
i don't know if `__xxx_TYPE__` is available there.

anyway, if we decide particular boards are always compiled with a particular compiler, we can simplify them later.
i don't feel it needs to be done in this PR.

Reply:
Sure, let's focus on clear the printf warning in this PR.